[MRG+2] Graph lasso tests

bnaul opened the pull request:
Added/improved some GraphLasso tests, per #4134.

1) Hard-coded optimal values for a well-known dataset.
2) Test 'lars' method as well as 'cd'.
3) Fix warning in `test_graph_lasso_cv` due to badly chosen warm starts.

agramfort commented:
travis is not happy.

also run a pep8 checker on your files.

bnaul commented:
Interesting...tests pass for me on OS X but not on Linux; I'm looking into it. And thanks for the PEP8 heads up.

bnaul commented:
Travis has been placated, think everything should be fine now.

agramfort commented on sklearn/covariance/tests/test_graph_lasso.py:
```diff
@@ -26,21 +28,24 @@ def test_graph_lasso(random_state=0):
             # Check that the costs always decrease (doesn't hold if alpha == 0)
             if not alpha == 0:
-                assert_array_less(np.diff(costs), 0)
+                assert_array_less(np.diff(costs), 1e-3)
```
why did you need to change this? this checks that the cost actually goes down at each iteration.

bnaul replied:
For badly-scaled problems the change in cost can be slightly positive, on the order of say 1e-12...that was causing failures for some test cases I'd tried, but now I guess it would be fine to make it 0, or maybe 1000*machine epsilon?

agramfort commented on sklearn/covariance/tests/test_graph_lasso.py:
```diff
@@ -69,9 +127,9 @@ def test_graph_lasso_cv(random_state=1):
         sys.stdout = StringIO()
         # We need verbose very high so that Parallel prints on stdout
-        GraphLassoCV(verbose=100, alphas=3, tol=1e-1).fit(X)
+        GraphLassoCV(verbose=100, alphas=5, tol=tol).fit(X)
```
why did you need to change this?

bnaul replied:
The alphas change is mentioned in the commit message: for widely-spaced alphas, the warm starts used by the CV loop are very bad, which actually can lead to lack of convergence (see http://web.stanford.edu/~hastie/Papers/glassoinsights.pdf).

Tolerance: the tolerance parameter is funny because the same value is used for both the overall glasso algorithm and for the elastic net solver inside the main loop. So even if we are OK with a high tolerance for Theta, passing one in here can cause the algorithm to fail to converge because the (lasso) steps are very far from optimal.

amueller replied:
So should we change the tol parameter and add a separate tol for the enet?

agramfort replied:
try to minimize the diff and I'll get another look.

bnaul replied:
My preference would be to add a separate tolerance for the enet, but I was a bit hesitant since I'm new to the project. If you guys agree that it makes sense to separate them then I will make the change; I can certainly imagine wanting to change the graph_lasso tolerance but don't see why anyone would ever want to mess with the enet precision.

agramfort commented on sklearn/covariance/tests/test_graph_lasso.py:
```diff
@@ -69,9 +127,9 @@ def test_graph_lasso_cv(random_state=1):
 
     # Smoke test with specified alphas
-    GraphLassoCV(alphas=[0.8, 0.5], tol=1e-1, n_jobs=1).fit(X)
+    GraphLassoCV(alphas=[0.8, 0.5], tol=tol, n_jobs=1).fit(X)
```
why did you need to change this?

bnaul replied:
See above.

agramfort commented:
LGTM

@GaelVaroquaux any objection?

ogrisel commented on sklearn/covariance/graph_lasso_.py:
```diff
@@ -79,8 +79,9 @@ def alpha_max(emp_cov):
-                max_iter=100, verbose=False, return_costs=False,
-                eps=np.finfo(np.float).eps, return_n_iter=False):
+                enet_tol=1e-4, max_iter=100, verbose=False,
+                return_costs=False, eps=np.finfo(np.float).eps,
```
It's better to be explicit with the precision: `eps=np.finfo(np.float64).eps`. We don't want platform specific behaviors.

bnaul replied:
That wasn't something I changed but I agree; I made the switch here and also in one other place (line 594).

ogrisel commented on sklearn/covariance/tests/test_graph_lasso.py:
```diff
@@ -69,7 +125,7 @@ def test_graph_lasso_cv(random_state=1):
         sys.stdout = StringIO()
         # We need verbose very high so that Parallel prints on stdout
-        GraphLassoCV(verbose=100, alphas=3, tol=1e-1).fit(X)
+        GraphLassoCV(verbose=100, alphas=5, tol=1e-1).fit(X)
```
why the change here?

bnaul replied:
See https://github.com/scikit-learn/scikit-learn/pull/4613#discussion_r28698920.

ogrisel commented:
LGTM.

ogrisel commented:
@bnaul could you please squash those commits together. There are many intermediate commits that are not interesting individually.

bnaul committed:
Added/improved GraphLasso tests, per #4134

1) Hard-coded optimal values for a well-known dataset.
2) Test 'lars' method as well as 'cd'.
3) Fix warning in test_graph_lasso_cv due to badly chosen warm starts.
4) Added separate tolerance parameter for enet solver (for method='cd').

bnaul commented:
Squashed!

ogrisel commented:
Merging. Thanks @bnaul !

amueller commented:
Shouldn't the ``enet_tol`` be exposed to the ``GraphLasso`` class?

bnaul commented:
Good point! I added the parameter to the other relevant places in GraphLasso here: 0cc36b6a1c7ebaf69ffdf8e9940c0d65488ab060. Would the proper procedure be to make another pull request, or is there some other way for a patch like this?

amueller commented:
please create a nother pull-request.
